Add a subset relation defined over lists

HuStmpHrrr opened the pull request:
as per #292, this PR attempts to implement a portion to illustrate the idea.

MatthewDaggitt commented:
Thanks for the PR! I'm not quite sure this was quite what I had in mind though. I agree we should have a set interface for `List`, but in my mind we shouldn't have the `subset` relation in `List.Relation` as it's no different from the current `sublist` relation. What does matter is the type of equality it's used with right?

I had understood the plan to be to add a new equality type `l1 ~~ l2 = l1 ⊆ l2 x l2 ⊆ l1` in `Data.List.Relation.Equality`?

The existing sublist relation then becomes a subset relation when parameterised by that equality?

A few of your lemmas already exist:
- `All-drop` lemmas already exist, see `++ˡ` etc. further up the file.
- `∈-concat⁻`s lemmas already exist in `Data.List.Membership.Setoid.Properties` (and they don't seem to have anything particularly to do with the `subset/list` relation?)

HuStmpHrrr commented:
@MatthewDaggitt  right. it can also be considered in that way.

I was thinking in another way after I opened this PR. As I mentioned in the issue, this is more general than it is shown in this PR as the equivalence relation itself has nothing to do with subset/sublist relation, but just the the fact that `_⊆_` is both transitive and reflexive would be sufficient to induce the said relation.

I will update this PR later to reflect this thought. I will also try to refactor this portion to `Data.List.Relation.Equality` as well.

MatthewDaggitt commented:
> As I mentioned in the issue, this is more general than it is shown in this PR as the equivalence relation itself has nothing to do with subset/sublist relation, but just the the fact that _⊆_ is both transitive and reflexive would be sufficient to induce the said relation.

Yup. I think we should have the definition of the new equality explicitly written out to make it clear to users what is going on, and then yes as you say we should generate the proofs for the equality relation via the `InducedEquivalence` proofs you've been referencing.

We can then add a `Data.List.Sets` file (similar to `Data.AVL.Sets`) that uses the above to provide all the right definitions to implement a finite set interface.

Does that sound like a reasonable plan?

HuStmpHrrr commented:
@MatthewDaggitt  and I'd imagine `Data.List.Sets` includes two submodules for two different equalities? How do you suggest to organize equalities?

HuStmpHrrr commented:
Sorry for delaying so much. there is always something during the final period.

I've implemented `Data.Lists.Sets` modules in a way that the differences are postponed at the point of specification of equivalence over Lists.

For subset relation, more general induced structures are defined and instantiated to obtain the previously agreed equivalence relation (+ more).

Sublist modules are also removed and dependencies are transited. 

@MatthewDaggitt Please kindly review if you have time and see if it's OK to implement in such a way.

MatthewDaggitt commented:
Hi @HuStmpHrrr. Sorry for taking a little while to reply. 

My apologies, again I'm afraid this isn't quite what I had in mind. I think it's definitely going to be cleaner if all relations are defined primarily with lists in mind, i.e. we keep `Data.List.Relation.Sublist`. The `Set` hierarchy should be very shallow and not contain any relation definitions. Instead it should import and rename the existing relations from the main `Data.List.Relation` hierarchy. As for dealing with the different types of equality I would think something like `Data.List.Relation.Set.X` where `X` is one of `Propositional\Setoid\Equivalence`.

I think there's lots of good ideas in here. They just need to be submitted in much smaller chunks! For instance a PR that only added the induced structures in your `Relation.Binary.InducedStructures` to `Relation.Binary.Properties.Preorder` would be quickly accepted. Likewise your membership reasoning module is really cool and would definitely be added on its own. 

In terms of the main set interfaces, I would recommending submitting a PR with that pretty much only contains `Data.List.Set.Setoid` to start with. We can then work through that in isolation.

HuStmpHrrr commented:
@MatthewDaggitt alright i will break it down into smaller pieces.

but before I close this PR, regarding the induced structure, I know that it's in fact induced from preorder, but the problem is, within this library, a preorder relation itself ships with an underlying equivalence relation. this makes preorder relation is not merely transitivity + reflexivity. I definitely want to distinguish these two. what would be your suggestion?

MatthewDaggitt commented:
> but before I close this PR, regarding the induced structure, I know that it's in fact induced from preorder, but the problem is, within this library, a preorder relation itself ships with an underlying equivalence relation. this makes preorder relation is not merely transitivity + reflexivity. I definitely want to distinguish these two. what would be your suggestion?

Sorry, missed this question somehow. I hope #333 has resolved it.
